### ISSUE TITLE
Mini-NAP: Adopt a version of NEP 23 – Backwards compatibility and deprecation policy

### ISSUE THREAD

**Issue by Carreau** [Napari/napari#4758]
It would be good to adopt a policy on Implementing deprecations and removals in napari. 
I found that a number of deprecation warnings are sub-par and might not be actionable, 

Plus it would be good to have a release step process to check that the deprecations messages are correct and that any promises of removal are enacted or updated when removal is postponed.

From Nep 23 and emphasis to section which _I believe_ we should adapt:

---- 

Deprecations:

 - shall include the version number of the release in which the functionality was deprecated.

 - shall include information on alternatives to the deprecated functionality, or a reason for the deprecation if no clear alternative is available. Note that release notes can include longer messages if needed.

 - shall use **DeprecationWarning by default, and VisibleDeprecation for changes that need attention again after already having been deprecated or needing extra attention for some reason**.

 - shall be listed in the release notes of the release where the deprecation is first present.

 - **shall not be introduced in micro (bug fix) releases**.

 - shall set a stacklevel, so the warning appears to come from the correct place.

 - shall be mentioned in the documentation for the functionality. A `.. deprecated::` directive can be used for this.

--- 

Obviously the `DeprecationWarning by default, and VisibleDeprecation` need to be discussed, as currently we have a mix of FutureWarnings, and VisibleDeprecation is not a thing in napari. (https://github.com/napari/napari/issues/4731) 

And as we mostly have patch release currently I think that deprecation only in minor release might be not aggressive enough.

There is some discussions in #4756 as well.

**Comment by Czaki** [comment]
Support

**Comment by tlambert03** [comment]
yeah this would be great.  thanks @Carreau.

> shall not be introduced in micro (bug fix) releases.

this one point works fine for numpy, given that they've freed themselves from https://0ver.org/ ... and have something much closer to honoring true semver, but we'd need to make some changes to our versioning philosophy for that to work here

**Comment by Carreau** [comment]
> this one point works fine for numpy, given that they've freed themselves from https://0ver.org/ ... and have something much closer to honoring true semver, but we'd need to make some changes to our versioning philosophy for that to work here

Yeah, that's why I said it's one of the point we might want to change. 
We could also decide to increase minor versions number more often.

**Comment by jni** [comment]
Yeah, I'm ok with moving faster in the minor versions, and ending up in 0.99 if need be. I'm also very much in favour of being generally more aggressive in napari regarding deprecations. Even though it is a popular library, we still bill ourselves as alpha, and there have been many instances where a deprecation pathway has been >2x the effort of just changing a thing. So I'd like to keep the option of "just" changing APIs between 0.(n) and 0.(n+1), as long as we have documentation to inform users of how to transition.

All this should indeed be codified in a NAP, regular size even. 😂 Thanks @Carreau!
